File: typescript/Leetcode_00108ConvertSortedArraytoBinarySearchTree.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function sortedArrayToBST(nums: number[]): TreeNode | null {
  return createNode(nums, 0, nums.length);
}

function createNode(
  nums: number[],
  left: number,
  right: number
): TreeNode | null {
  if (left <= right) {
    const mid = Math.floor((left + right) / 2);
    const root = new TreeNode(
      nums[mid],
      createNode(nums, left, mid - 1),
      createNode(nums, mid + 1, right)
    );
    return root;
  }
  return null;
}

sortedArrayToBST([-10,-3,0,5,9]);

File: java/src/main/java/tech/datvu/leetcode/general/Leetcode_00108ConvertSortedArraytoBinarySearchTree.java
package tech.datvu.leetcode.general;

import tech.datvu.leetcode.shared.global.TreeNode;

public class Leetcode_00108ConvertSortedArraytoBinarySearchTree {
    public static TreeNode sortedArrayToBST(int[] nums) {
        return insert(nums, 0, nums.length - 1);
    }

    private static TreeNode insert(int[] nums, int left, int right) {

        // Neu mang khong ton tai
        if (left > right) {
            return null;
        }

        // Neu mang chi co 1 phan tu
        if (left == right) {
            return new TreeNode(nums[left]);
        }

        // Dat root bang gia tri o giua
        int mid = (left + right) / 2;
        TreeNode root = new TreeNode(nums[mid]);

        /* Bai toan con cua bai toan lon */
        // Gan ve trai bang gia tri o ben trai mang
        root.left = insert(nums, left, mid - 1);
        // Gan ve phai bang gia tri o ben phai mang
        root.right = insert(nums, mid + 1, right);
        return root;
    }

    public static void main(String[] args) {
        int[] nums = { -10, -3, 2, 3, 4, 5 };
        TreeNode res = sortedArrayToBST(nums);
        TreeNode.preTraversalNode(res);
        System.out.println();
    }
}
